File: ratdb/End/display_gdml.py
import sys
import subprocess
import os

def display_gdml_in_root(gdml_file_path):
    """
    Displays a GDML file in ROOT.

    Args:
        gdml_file_path (str): The absolute path to the GDML file.
    """
    if not os.path.exists(gdml_file_path):
        print(f"Error: GDML file not found at {gdml_file_path}")
        return

    # ROOT commands to import and draw the GDML file
    root_commands = f"""
    TGeoManager::Import("{gdml_file_path}");
    if (gGeoManager) {{
        gGeoManager->GetTopVolume()->Draw();
        gApplication->Run(); // Keep the canvas open
    }} else {{
        std::cerr << "Error: TGeoManager not initialized." << std::endl;
    }}
    .q
    """

    print(f"Attempting to display {gdml_file_path} in ROOT...")
    try:
        # Run ROOT in batch mode and execute the commands
        process = subprocess.run(
            ["root", "-l", "-b"],  # -l: no ROOT splash screen, -b: batch mode
            input=root_commands,
            text=True,
            check=True,
            capture_output=True
        )
        print("ROOT Stdout:")
        print(process.stdout)
        print("ROOT Stderr:")
        print(process.stderr)
        print(f"Successfully attempted to display {gdml_file_path} in ROOT.")
    except subprocess.CalledProcessError as e:
        print(f"Error running ROOT: {e}")
        print(f"ROOT Stdout: {e.stdout}")
        print(f"ROOT Stderr: {e.stderr}")
    except FileNotFoundError:
        print("Error: 'root' command not found. Please ensure ROOT is installed and in your PATH.")

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python3 display_gdml.py <path_to_gdml_file>")
        sys.exit(1)

    gdml_file = sys.argv[1]
    display_gdml_in_root(gdml_file)

File: ratdb/End/display_gdml.cpp
#include "TGeoManager.h"
#include "TApplication.h"
#include "TCanvas.h"
#include "TFile.h"
#include "TSystem.h"
#include <iostream>
#include <string>

void display_gdml(const std::string& gdml_file_path) {
    // Check if the file exists
    if (gSystem->AccessPathName(gdml_file_path.c_str())) {
        std::cerr << "Error: GDML file not found at " << gdml_file_path << std::endl;
        return;
    }

    // Import the GDML file
    TGeoManager::Import(gdml_file_path.c_str());

    if (gGeoManager) {
        // Create a canvas to draw on
        TCanvas *c1 = new TCanvas("c1", "GDML Viewer", 800, 600);
        c1->cd();

        // Draw the top volume
        gGeoManager->GetTopVolume()->Draw();

        // Update the canvas and run the ROOT application to keep it open
        c1->Update();
        gApplication->Run();
    } else {
        std::cerr << "Error: TGeoManager not initialized after import." << std::endl;
    }
}

int main(int argc, char** argv) {
    // Initialize ROOT application
    TApplication app("GDMLViewer", &argc, argv);

    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <path_to_gdml_file>" << std::endl;
        return 1;
    }

    std::string gdml_file = argv[1];
    display_gdml(gdml_file);

    return 0;
}